framework/rest/DataFilter.php

```diff
@@ -0,0 +1,809 @@
+     */
+    public $multiValueOperators = [
+        'IN',
+        'NOT IN',
+    ];
+    /**
+     * @var array actual attribute names to be used in searched condition, in format: [filterAttribute => actualAttribute].
+     * For example, in case of using table joins in the search query, attribute map may look like following:
+     *
+     * ```php
+     * [
+     *     'authorName' => '{{author}}.[[name]]'
+     * ]
+     * ```
+     *
+     * Attribute map will be applied to filter condition at [[normalize()]].
+     */
+    public $attributeMap = [];
+
+    /**
+     * @var array list of error messages responding to invalid filter structure, in format: `[errorKey => message]`.
+     */
+    private $_errorMessages = [
+        'invalidFilter' => 'The format of {filter} is invalid.',
+        'operatorRequireMultipleOperands' => "Operator '{operator}' requires multiple operands.",
+        'unknownAttribute' => "Unknown filter attribute '{attribute}'",
+        'invalidAttributeValueFormat' => "Condition for '{attribute}' should be either a value or valid operator specification.",
+        'operatorRequireAttribute' => "Operator '{operator}' must be used with search attribute.",
+        'unsupportedOperatorType' => "'{attribute}' does not support operator '{operator}'.",
+    ];
+    /**
+     * @var mixed raw filter specification.
+     */
+    private $_filter;
+    /**
+     * @var Model|array|string|callable model to be used for filter attributes validation.
+     */
+    private $_searchModel;
+    /**
+     * @var array list of search attribute types in format: attributeName => type
+     */
+    private $_searchAttributeTypes;
+
+
+    /**
+     * @return mixed raw filter value.
+     */
+    public function getFilter()
+    {
+        return $this->_filter;
+    }
+
+    /**
+     * @param mixed $filter raw filter value.
+     */
+    public function setFilter($filter)
+    {
+        $this->_filter = $filter;
+    }
+
+    /**
+     * @return Model model instance.
+     * @throws InvalidConfigException on invalid configuration.
+     */
+    public function getSearchModel()
+    {
+        if (!is_object($this->_searchModel) || $this->_searchModel instanceof \Closure) {
+            $model = Yii::createObject($this->_searchModel);
+            if (!$model instanceof Model) {
+                throw new InvalidConfigException('`' . get_class($this) . '::$model` should be an instance of `' . Model::className() . '` or its DI compatible configuration.');
+            }
+            $this->_searchModel = $model;
+        }
+        return $this->_searchModel;
+    }
+
+    /**
+     * @param Model|array|string|callable $model model instance or its DI compatible configuration.
+     * @throws InvalidConfigException on invalid configuration.
+     */
+    public function setSearchModel($model)
+    {
+        if (is_object($model)) {
+            if (!$model instanceof Model && !$model instanceof \Closure) {
```

Comment:
Could be moved to the condition above.

Reply:
No, it could not: `$model` may be an array or string class name, which is a valid value and should not throw an exception.

Reply:
> $model may be an array

Then it will not be object and no exception will be thrown

Reply:
fixed